File: daemon/src/ws/handlers/subscribe.ts
import type { WSConnection } from "../connection.js";
import type { ClientMessage } from "../protocol.js";

/**
 * Handle subscribe message: add session IDs to the subscription set.
 * The connection will then receive state events for these sessions via the broadcaster.
 */
export function handleSubscribe(conn: WSConnection, msg: Extract<ClientMessage, { type: "subscribe" }>): void {
  conn.subscribe(msg.sessionIds);
}

/**
 * Handle unsubscribe message: remove session IDs from the subscription set.
 */
export function handleUnsubscribe(conn: WSConnection, msg: Extract<ClientMessage, { type: "unsubscribe" }>): void {
  conn.unsubscribe(msg.sessionIds);
}

File: rust/vst-ws/src/handlers/subscribe.rs
//! `subscribe` / `unsubscribe` handlers.

use vst_types::ws::ClientMessage;

use crate::connection::WsConnection;

pub fn handle_subscribe(conn: &WsConnection, msg: &ClientMessage) {
    if let ClientMessage::Subscribe { session_ids } = msg {
        conn.subscribe(session_ids);
    }
}

pub fn handle_unsubscribe(conn: &WsConnection, msg: &ClientMessage) {
    if let ClientMessage::Unsubscribe { session_ids } = msg {
        conn.unsubscribe(session_ids);
    }
}
